service/storage_service.cc

```diff
@@ -834,6 +834,9 @@ future<> storage_service::handle_state_bootstrap(inet_address endpoint) {
 }
 
 future<> storage_service::handle_state_normal(inet_address endpoint) {
+    auto work = defer([this, endpoint] {
```

Comment:
Shouldn't we do this only if `handle_state_normal` succeeds?

Reply:
We do not allow handle_state_normal to fail, if it fails the node will abort to prevent it stays in a bad state forever.

Reply:
> We do not allow handle_state_normal to fail, if it fails the node will abort to prevent it stays in a bad state forever.

I don't think that's true. From looking at the code, if `handle_state_normal` throws the exception will be catched in gossiper code and a warning will be logged, then we'll retry. There's no code which aborts on exception.

Reply:
The abort on failure property was removed accidentally by Pavel Solodovnikov at c48dcf607a38fdf481cc161cf428b6c3002963f0. The code was running under noexcept.

```
-    // Exceptions thrown from listeners will result in abort because that could leave the node in a bad
-    // state indefinitely. Unless the value changes again, we wouldn't retry notifications.
-    // Some values are set only once, so listeners would never be re-run.
-    // Listeners should decide which failures are non-fatal and swallow them.
-    auto run_listeners = seastar::defer([&] () noexcept {
-        for (auto&& key : changed) {
-            do_on_change_notifications(addr, key, remote_map.at(key)).get();
+    std::exception_ptr ep;
+    try {
+        // we need to make two loops here, one to apply, then another to notify,
+        // this way all states in an update are present and current when the notifications are received
+        for (const auto& remote_entry : remote_map) {
+            const auto& remote_key = remote_entry.first;
+            const auto& remote_value = remote_entry.second;
+            auto remote_gen = remote_state.get_heart_beat_state().get_generation();
+            auto local_gen = local_state.get_heart_beat_state().get_generation();
+            if(remote_gen != local_gen) {
+                auto err = format("Remote generation {:d} != local generation {:d}", remote_gen, local_gen);
+                logger.warn("{}", err);
+                throw std::runtime_error(err);
+            }
+
+            const versioned_value* local_val = local_state.get_application_state_ptr(remote_key);
```

Reply:
I will drop the defer usage for now.